delta/plugins/archive/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/archive/ArchiveDownload.scala

```diff
@@ -0,0 +1,189 @@
+        case ResourceRef.Latest(iri)           => files.fetchContent(iri, refProject)
+        case ResourceRef.Revision(_, iri, rev) => files.fetchContentAt(iri, refProject, rev)
+        case ResourceRef.Tag(_, iri, tag)      => files.fetchContentBy(iri, refProject, tag)
+      }
+      val tarEntryIO     = fileResponseIO
+        .map { fileResponse =>
+          val path     = ref.path.map(_.value.toString).getOrElse(pathOf(ref, project, fileResponse.filename))
+          val metadata = TarArchiveMetadata.create(path, fileResponse.bytes)
+          Some((metadata, fileResponse.content))
+        }
+        .mapError {
+          case _: FileRejection.FileNotFound                 => ResourceNotFound(ref.ref, project)
+          case _: FileRejection.TagNotFound                  => ResourceNotFound(ref.ref, project)
+          case _: FileRejection.RevisionNotFound             => ResourceNotFound(ref.ref, project)
+          case FileRejection.AuthorizationFailed(addr, perm) => AuthorizationFailed(addr, perm)
+          case other                                         => WrappedFileRejection(other)
+        }
+      if (ignoreNotFound) tarEntryIO.onErrorRecover { case _: ResourceNotFound => None }
+      else tarEntryIO
+    }
+
+    private def fetchResource(
+        ref: ResourceReference,
+        project: ProjectRef,
+        ignoreNotFound: Boolean
+    ): IO[ArchiveRejection, Option[(TarArchiveMetadata, AkkaSource)]] = {
+      val tarEntryIO = resourceRefToByteString(ref, project).map { content =>
+        val path     = ref.path.map(_.value.toString).getOrElse(pathOf(ref, project))
+        val metadata = TarArchiveMetadata.create(path, content.length.toLong)
+        Some((metadata, Source.single(content)))
+      }
+      if (ignoreNotFound) tarEntryIO.onErrorHandle { _: ResourceNotFound => None }
+      else tarEntryIO
+    }
+
+    private def resourceRefToByteString(
+        ref: ResourceReference,
+        project: ProjectRef
+    ): IO[ResourceNotFound, ByteString] = {
+      val p = ref.project.getOrElse(project)
+      val r = ref.representation.getOrElse(CompactedJsonLd)
+      UIO
+        .tailRecM(exchanges.toList) { // try all reference exchanges one at a time until there's a result
+          case Nil        => UIO.pure(Right(None))
+          case ex :: rest => ex(p, ref.ref).map(_.toRight(rest).map(value => Option(value)))
+        }
+        .flatMap {
+          case Some(value) => valueToByteString(value, r).logAndDiscardErrors("serialize resource to ByteString")
+          case None        => IO.raiseError(ResourceNotFound(ref.ref, project))
+        }
+    }
+
+    private def valueToByteString[A](
+        value: ReferenceExchangeValue[A],
+        repr: ArchiveResourceRepresentation
+    ): IO[RdfError, ByteString] = {
+      implicit val encoder: JsonLdEncoder[A] = value.encoder
+      repr match {
+        case SourceJson      => UIO.pure(ByteString(value.toSource.sort.spaces2))
```

Comment:
We could have a method prettyPrint or something that would do this `.sort.spaces2` ?
Or maybe it is not that important ?

Reply:
I've already commented on how https://github.com/BlueBrain/nexus/pull/2114#discussion_r592232591